### PR TITLE
Adding omsl - OpenModelica Modelica Standard Language

### PR DESCRIPTION
**Pull request by joewa**
Adding Modelica Standard Language that can be used with the package omcompiler

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by joewa** [comment]
I am happy to provide my first PR and I suppose its ready for a review.

Can you please have a look @conda-forge/staged-recipes or `@conda-forge/help-python-c`

Many thanks!

File: recipes/omsl/conda_build_config.yaml
Comment (carterbox) [inline recipes/omsl/conda_build_config.yaml:3]: Do not use the conda_build_config to build multiple versions of osml. Start with the older version, then once this PR is merged, open a PR on the feedstock to update to the newer version. If you need to maintain v3.x and v4.x simultaneously, then create a branch for each in the feedstock. Every commit on the feedstock is published.

File: recipes/omsl/meta.yaml
Comment (carterbox) [inline recipes/omsl/meta.yaml:20]: Why is python listed in the host requirements, the install script is only copying files, it doesn't look like python is used at all?

**Review by carterbox — CHANGES_REQUESTED** [review]
@conda-forge/omcompiler, I'm not familiar with this Modelica language. Is it standard practice to install uncompiled libraries to `/lib/omlibrary/`? Are `mo` files compiled or are they interpreted like python files?

File: recipes/omsl/meta.yaml
Comment (joewa) [reply recipes/omsl/meta.yaml]: Agreed, I will reach out to the maintainer of @conda-forge/omcompiler-feedstock asking him to add these packages in the run section.
These packages are installed on most many hosts by default. I found that they are required when installing omcompiler on a fresh container in the cloud. Can you please have a look @jschueller ?

In the first step, `omcompiler` builds C/C++ code from the mo-files (and the libraries). In the 2nd step, it calls gxx to build a binary that represents the model and returns the timeseries. The the "Modelica Standard Library" is used as a foundation for almost all Modelica models. This feedstock is intended to make it  installable through conda.

And yes, `omcompiler` expects to find the Modelica libraries either in `/lib/omlibrary/` or in `~/.openmodelica/libraries/`. The 1st is the default, e.g. when the omcompiler and the Modelica libraries are installed though `apt-get` in a Debian environment (of course, `conda` and `apt` will never be mixed). The 2nd is used by the [package manager](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/packagemanager.html) which is built in `omcompiler` enabling the installation of Modelica libraries in a mo-script via `installPackage`. Unfortunately, this cannot be used in our CI environment.

File: recipes/omsl/meta.yaml
Comment (joewa) [reply recipes/omsl/meta.yaml:20]: Correct, I will remove it.

File: recipes/omsl/conda_build_config.yaml
Comment (joewa) [reply recipes/omsl/conda_build_config.yaml]: OK, understood, thanks for explaining this. Indeed, I will have to maintain v3.x and v4.x because they are not compatible and v3.x is still widely used in production. So I removed conda_build_config.

**Comment by joewa** [comment]
I hope its OK now.
Can you please have a look @conda-forge/staged-recipes @carterbox ?
Many thanks!

File: recipes/omsl/meta.yaml
Comment (carterbox) [reply recipes/omsl/meta.yaml]: Why aren't the mo-files precompiled and shipped and dynamically linked into end-users' models? Is this a JIT compiler-type framework where mo-files are compiled as-needed?

File: recipes/omsl/meta.yaml
Comment (carterbox) [reply recipes/omsl/meta.yaml]: Same question for pip. Why is pip needed to install these libraries?

File: recipes/omsl/meta.yaml
Comment (carterbox) [inline recipes/omsl/meta.yaml:22]:
```suggestion
```
omsl is a library right? It should not depend on a compiler package at runtime. Users should know that they need a compiler to use these files. It seems that this library is very similar to a C/CXX header-only library i.e. Eigen. The Eigen package is a collection of source files, and it does not depend on a compiler.

**Review by carterbox — CHANGES_REQUESTED** [review]

**Review by carterbox — CHANGES_REQUESTED** [review]
If this package is really just a source distribution, then we could probably mark it as `noarch: generic` as long as the install directory is the same for linux/unix/windows?

File: recipes/omsl/meta.yaml
Comment (jschueller) [reply recipes/omsl/meta.yaml]: feel free to send a PR against the omcompiler package

Comment (joewa) [reply recipes/omsl/meta.yaml]: This is indeed a JIT compiler-type framework where the mo-files are compiled as needed. In general "a model" consists of a set of (mostly) acausal "differential algebraic equations" (DAE) where the number of variables and equations is always equal. Most user models are composed from (sub-) models from the MSL (this library). The mo-files cannot be precompiled because the causality (i.e. the direction of a signal flow) depends on how a model is composed (as far as I understood). omcompiler creates the C code which has causality.

**Comment by joewa** [comment]
I agree, we could mark this package (omsl) as `noarch: generic` because this is really just a source distribution. But please mind that its probably useless without `omcompiler`, which is currently available for Linux only. There might be a Windows release in the future but [MaxOSX](https://openmodelica.org/download/download-mac) will never be supported.

The current tests are invoking `omcompiler` to build a model from this library which ensures that the library is installed in the correct location and OK.
So if we would mark it as `noarch: generic`, I suppose we would have to remove the tests because `omcompiler` is available for Linux only.

My proposal would be to keep it as it is. Of course `noarch: generic` will work, but I will have to remove the tests then.

What do you think @carterbox ?

**Comment by joewa** [comment]
Can you please have a look @conda-forge/staged-recipes @carterbox ?

**Review by carterbox — APPROVED** [review]
Thanks for answering all my questions. Consider noarch:generic if future releases of the compiler are available for other platforms.

**Comment by joewa** [comment]
Thanks you too for all your explanations!
I learned a lot!

**Pull request merged**
